### PR TITLE
fix(core): Eliminate false-negative 500s in workflow builder tools

### LINKED ISSUE #31328: [MCP SDK] create_workflow_from_code intermittently returns HTTP 500, often as a false negative (workflow persists anyway, causing duplicates on retry)
## Summary

The native n8n Workflow Builder MCP server's `create_workflow_from_code` **intermittently** returns HTTP 500 for valid code that `validate_workflow` accepts. Three things distinguish this from the existing deterministic 500 reports (e.g. #29660):

1. **Non-deterministic**: the *same* code, submitted twice in a row, can 500 once and succeed once.
2. **Not node-specific**: observed with `n8n-nodes-base.emailReadImap`, and separately with `n8n-nodes-base.gmail` (v2.2) + `n8n-nodes-base.scheduleTrigger` (no IMAP node involved).
3. The 500 is frequently a **false negative**: the workflow is persisted in n8n despite the error returned to the client, so a retry creates a **duplicate**.

## Environment

- n8n: self-hosted (Docker), version **2.18.5**
- Builder MCP / `@n8n/workflow-sdk`
- Client: Claude Code MCP integration
- Observed: 2026-05-26 and 2026-05-27

## Observed behavior

**Intermittency.** Two independent MCP sessions submitted near-identical code containing an `n8n-nodes-base.emailReadImap` trigger. One returned HTTP 500; the other, minutes later, succeeded with the same node, with no config or credential difference that would explain it.

**Not node-specific.** The next day the same 500 hit a workflow with no IMAP node at all (`n8n-nodes-base.gmail` v2.2 + `n8n-nodes-base.scheduleTrigger`), so the fault is in the tool/persist path, not a specific node's schema.

**False negative -> duplicates (the most actionable part).** `create_workflow_from_code` returned 500 to the client, but the workflow was created in n8n anyway. Retrying after the 500 produced a second, identical workflow. In one case two consecutive 500s left two complete, identical workflows that had to be reconciled by hand (`search_workflows` to detect, `archive_workflow` to clean up).

**No error body.** As in #29660, the 500 carries no error payload to the MCP client, so the underlying cause can't be diagnosed client-side.

## Expected behavior

- When `validate_workflow` accepts the code, `create_workflow_from_code` should persist deterministically.
- A genuine failure should not leave a persisted workflow (no partial commit), and/or the call should be idempotent or return the created workflow id so retries don't duplicate.
- The actual error should be surfaced in the response body.

## Impact

Workflow creation via the Builder MCP is unreliable and silently lossy toward duplicates: every 500 forces a defensive "`search_workflows` before retrying" step, and missed duplicates accumulate in the instance. This undercuts the core value of building workflows programmatically.

## Suggested triage

The combination of (a) persist-succeeds-but-500-returned and (b) non-determinism points at the **post-persist / response-construction** step throwing after the DB write has committed (e.g. a serialization step, credential auto-assignment, or a race/timeout), rather than the create itself failing. Two concrete asks:

1. Surface the real error body to the MCP client; this alone would let reporters file deterministic repros.
2. Make create persist+respond atomic, or idempotent / id-returning, so a 500 can't leave an orphan duplicate.

## Repro notes

The bug is non-deterministic, so there's no single deterministic snippet; it reproduces by calling `create_workflow_from_code` repeatedly with the same valid, `validate_workflow`-passing code and observing occasional 500s plus duplicate persists. Happy to share the exact SDK code and, if you point me at the relevant server-side logger, capture a stack trace from a self-hosted instance.

Related: #29660 (deterministic 500 + no error body on `create_workflow_from_code` for langchain root nodes); likely shares the response-path handling.

### PR DESCRIPTION
**Pull request by sinehypernova-0718**
## Summary

Fixes intermittent false-negative 500 errors returned by the MCP workflow-builder tools `create_workflow_from_code` and `update_workflow`.

**Root cause:** Post-save side effects (telemetry, credential auto-assign tracking, hook execution) could throw exceptions after the database write had already been committed. Since the tools couldn't tell the difference between a real persistence failure and a post-save side-effect error, they would return `isError: true` to the MCP client, often leading to duplicate workflows when the client retried.

**What this PR does:**
- Isolates post-save side effects in both tools so that failures are logged and tracked with metrics, but never cause a successful database write to be returned as a client-visible 500.
- Adds a recovery fallback to `update_workflow` (matching the existing behavior in `create_workflow_from_code`). If an update throws, the tool re-fetches the workflow, verifies that the expected state was actually persisted (content match for graph changes, `updatedAt` freshness for settings-only changes), and returns success with a `note` if the write completed successfully.
- Strengthens the existing create-tool fallback by logging lookup failures that were previously silent and isolating telemetry in the recovery path.
- Adds a machine-readable `errorCode` to every error response (e.g. `HTTP_404`, `MISSING_PROJECT_ID`, `UNKNOWN_ERROR`).
- Introduces `McpPostSaveMetricsService` with a Prometheus counter `mcp_post_save_failures_total` (labels `tool`, `error_type`) for production observability.
- Ensures recovery logic only runs after a database update has actually been attempted, avoiding unnecessary lookups for validation errors.

> [!NOTE]
> Client-supplied idempotency keys with database-backed deduplication (I guess, creating an internal ticket for the same is enough for now, if the team decides so.)

### AI Disclosure

I used Claude Code with the Minimax:M3 model through the free tier of Ollama Cloud to find the relevant files and functions, related to the issue. Also, used it for finding edge cases and race-conditions which I missed initially. Used, GPT-5.5 through Codex for writing tests (test plan drafted by me). And, ofcourse AI tab autocompletions in the IDE :) CodeRabbit and Cubic for automated code review in the CI of my own fork. Also, added markdown formatting and fixed grammar and improved the wording of the PR description.

## How to test

1. **Unit tests** – `pnpm test src/modules/mcp/__tests__/update-workflow.tool.test.ts` and `create-workflow-from-code.tool.test.ts` cover all new behavior: side-effect isolation, recovery after post-save errors (graph and settings-only operations), genuine failures when the persisted state doesn't match, error-code mapping, and lookup failure logging. All existing tests continue to pass without modification.

2. **Integration**
   - Start a local n8n instance and connect an MCP client (e.g. Claude Code).
   - Call `create_workflow_from_code` with any valid workflow code.
   - Temporarily force a post-save side effect to throw (for example, patch `telemetry.track` to throw inside the isolated `try/catch`).
   - Verify that the client receives a successful response containing `workflowId`, `name`, and `url`, and that the workflow exists in the database.
   - Repeat the same steps for the update tool: update a workflow, force a post-save error, confirm the update was applied, and verify that the client receives a successful response with a `note`.

## Related Linear tickets, GitHub issues, and Community forum posts

- Fixes #31328
- Linear Ticket ID: [GHC-8540](https://linear.app/n8n/issue/GHC-8540)
- Related #29660

## Review / Merge checklist

- [x] I have seen this code, I have run this code, and I take responsibility for this code.
- [x] PR title and summary are descriptive. ([conventions](../blob/master/.github/pull_request_title_conventions.md)) 
- [ ] [Docs updated](https://github.com/n8n-io/n8n-docs) or follow-up ticket created.
- [x] Tests included. 
- [ ] PR Labeled with `Backport to Beta`, `Backport to Stable`, or `Backport to v1` (if the PR is an urgent fix that needs to be backported)

---

<a href="https://stagereview.app/n8n-io/n8n/pull/34975">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://stagereview.app/assets/gh-open-in-stage-dark.svg">
    Open in Stage
  </picture>
</a>

<a href="https://cubic.dev/pr/n8n-io/n8n/pull/34975?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

**Comment by sinehypernova-0718** [comment]
Hey! @garritfra Thanks for triggering the CI checks, I see the failures. I will update you once the changes are pushed.

**Comment by garritfra** [comment]
@sinehypernova-0718 thanks! And sorry for the label spam - we're testing out a new triaging system and this PR serves as a guinea pig 😄

**Comment by sinehypernova-0718** [comment]
Hii! The changes are pushed, `build`, `lint` and `typecheck` passes locally ✅. Also, thanks for routing it to the concerned team.

File: packages/cli/src/modules/mcp/mcp-post-save-metrics.service.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+	constructor(config: PrometheusMetricsConfig) {
+		this.config = config;
+	}
+
+	/**
+	 * Increment `mcp_post_save_failures_total{tool, error_type}`. `error_type`
+	 * is the class name of the thrown error, falling back to a stable string
+	 * for non-Error throws so the label cardinality stays bounded.
+	 */
+	incrementPostSaveFailure(tool: 'create' | 'update', error: unknown): void {
```
Comment (MiloradFilipovic) [inline packages/cli/src/modules/mcp/mcp-post-save-metrics.service.ts:19]: How about making this method swallow its own errors (wrap the body in try/catch + `logger.debug`)?
As documented, it's observability-only so this will prevent it from breaking a caller. That would also let you delete the nested try/catch around the calls in `update-workflow.tool.ts` and fix the unwrapped call in `create-workflow-from-code.tool.ts`.

File: packages/cli/src/modules/mcp/tools/workflow-builder/create-workflow-from-code.tool.ts [hunk 2]
```diff
@@ -467,27 +492,42 @@ export const createCreateWorkflowFromCodeTool = (
...
+						telemetry.track(USER_CALLED_MCP_TOOL_EVENT, telemetryPayload);
+					} catch (telemetryError) {
+						logger.error(
+							'Post-save telemetry failed for create_workflow_from_code (recovery path)',
+							{
+								workflowId: persisted.id,
+								error: telemetryError,
+							},
+						);
+						postSaveMetrics.incrementPostSaveFailure('create', telemetryError);
```
Comment (MiloradFilipovic) [inline packages/cli/src/modules/mcp/tools/workflow-builder/create-workflow-from-code.tool.ts]: As mentioned in the previous comment:
a) Error in post-save metrics should not break callers or
b) We need to wrap this call in `try`/`catch`

File: packages/cli/src/modules/mcp/tools/workflow-builder/update-workflow.tool.ts [hunk 3]
```diff
@@ -1355,20 +1404,160 @@ export const createUpdateWorkflowTool = (
...
+					const hasPersistedExpectedSettings =
+						expectedSettings !== undefined &&
+						persisted &&
+						isEqual(persisted.settings, expectedSettings) &&
+						!isEqual(existingWorkflow.settings, expectedSettings);
+					const wasTouched = Boolean(hasNewerTimestamp || hasPersistedExpectedSettings);
+					const tagsMatchExpected =
+						!hasTagOperations || haveSameTagNames(persisted?.tags, expectedTagNames);
+
+					const recovered =
```
Comment (MiloradFilipovic) [inline packages/cli/src/modules/mcp/tools/workflow-builder/update-workflow.tool.ts:1515]: If I read this correctly, there is a possible false success for no-op batches: with graph ops, `recovered` only requires `matchesExpected`. If the agent submits operations that leave the workflow identical to its stored state (e.g. a retry of an already-applied update), `expectedWorkflow` deep-equals the pre-update content - so a throw *inside* `workflowService.update` **before** the commit (pre-save hook rejection, tampering check) still reports "updated successfully, but a post-save operation failed". Since `updatedAt` is only persisted when `updateContent` runs, requiring the timestamp for the no-op case closes it:

```ts
const contentChanged =
	!isEqual(existingWorkflow.nodes, expectedWorkflow?.nodes) ||
	!isEqual(existingWorkflow.connections, expectedWorkflow?.connections);
const recovered =
	tagsMatchExpected &&
	(hasGraphOps ? Boolean(matchesExpected) && (contentChanged || hasNewerTimestamp) : wasTouched);
```

Looks like the settings path guards this with `!isEqual(existingWorkflow.settings, expectedSettings)`, but the graph path doesn't.

File: packages/cli/src/modules/mcp/tools/workflow-builder/update-workflow.tool.ts [hunk 4]
```diff
@@ -1355,20 +1404,160 @@ export const createUpdateWorkflowTool = (
...
+						logger.warn('Post-update verification lookup failed', {
+							workflowId,
+							error: lookupError,
+						});
+					}
+
+					const matchesExpected =
+						persisted &&
+						expectedWorkflow &&
+						isEqual(persisted.nodes, expectedWorkflow.nodes) &&
```
Comment (MiloradFilipovic) [inline packages/cli/src/modules/mcp/tools/workflow-builder/update-workflow.tool.ts]: Should we normalize both sides here before comparison? Something like:

```ts
const normalize = (value: unknown) => JSON.parse(JSON.stringify(value ?? null));
// isEqual(normalize(persisted.nodes), normalize(expectedWorkflow.nodes)) etc.
```

Context: Here, we compare an in-memory entity against a DB round-trip, and lodash treats `{ key: undefined }` and `{}` as different - so a node parameter explicitly set to `undefined` (or any non-JSON value like a `Date`) makes recovery miss a genuinely committed write and result in false negative.

File: packages/cli/src/modules/mcp/tools/workflow-builder/create-workflow-from-code.tool.ts [hunk 5]
```diff
@@ -226,6 +236,8 @@ export const createCreateWorkflowFromCodeTool = (
 			openWorldHint: false,
 		},
 	},
+
+	// eslint-disable-next-line complexity
```
Comment (MiloradFilipovic) [inline packages/cli/src/modules/mcp/tools/workflow-builder/create-workflow-from-code.tool.ts]: Instead of doing this, how about extracting the recovery block into a named helper (e.g. `recoverPersistedCreate(...)`)? The update tool was deliberately refactored into named steps, and the two tools now share the recovery/telemetry scaffolding.

File: packages/cli/src/modules/mcp/mcp.module.ts [hunk 6]
```diff
@@ -14,6 +14,8 @@ export class McpModule implements ModuleInterface {
 	async init() {
 		await import('./mcp.controller.js');
 		await import('./mcp.settings.controller.js');
+		const { McpPostSaveMetricsService } = await import('./mcp-post-save-metrics.service.js');
+		Container.get(McpPostSaveMetricsService);
```
Comment (MiloradFilipovic) [inline packages/cli/src/modules/mcp/mcp.module.ts]: Unless there's a registration-order reason I'm missing, this eager `Container.get(McpPostSaveMetricsService)` looks redundant. `McpService` constructor-injects the service, so DI constructs it on first use, and the counter itself is lazy anyway. Please double-check this but I think these two lines can go

File: packages/cli/src/modules/mcp/mcp-post-save-metrics.service.ts [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+		if (!this.config.enable) return;
+
+		this.postSaveFailuresTotal ??= new Counter({
```
Comment (MiloradFilipovic) [inline packages/cli/src/modules/mcp/mcp-post-save-metrics.service.ts]: Every other prom-client metric in `cli` lives in `packages/cli/src/metrics/prometheus/` as a `PrometheusMetricsCollector` (see `poll-trigger-metrics.service.ts`), wired into `PrometheusMetricsService`'s collectors list and usually gated behind an `include*Metrics` toggle. This is the first direct `new Counter(...)` outside that folder. Since a metric name is an operational contract, I'd prefer either moving this into the metrics module (e.g. the tools emit an event and a collector subscribes, like the existing collectors do).

File: packages/cli/src/modules/mcp/tools/workflow-builder/create-workflow-from-code.tool.ts (hunk 2 above)
Comment (MiloradFilipovic) [reply packages/cli/src/modules/mcp/tools/workflow-builder/create-workflow-from-code.tool.ts]: Also, should we call the `postSaveMetrics.incrementPostSaveFailure` in the create recovery branch (before returning recovered response) ti make sure recovered failuresare counted in this new metric?

**Review by MiloradFilipovic — COMMENTED** [review]
Hey @sinehypernova-0718 
Thanks for handling this and sorry for waiting on the review. I left a few feedback points in inline comments so take a look when you get a chance.

EDIT: Got me thinking if we are making the fix for this more involved than it should be? If the question we are trying to answer here is `were the changes committed?`, then maybe we can answer it in the `workflow.service`, probably with something like this, after `updateContent` returns:

```
const updated = await workflowService.update(user, data, id, {
  source: 'n8n-mcp',
  collectPostSaveError: true,  // only MCP passes this
});
// returns { workflow, postSaveError?: Error } shape for opted-in callers

// THEN ON THE TOOL SIDE: if (postSaveError) return successWithNote(...)
```

I know this will not solve the create tool but, if sound, it will establish a pattern that we can retrofit there. Let me know what you think.

**Comment by sinehypernova-0718** [comment]
Hi! @MiloradFilipovic The required changes are pushed now. ✅

> EDIT: Got me thinking if we are making the fix for this more involved than it should be? If the question we are trying to answer here is `were the changes committed?`, then maybe we can answer it in the `workflow.service`, probably with something like this, after `updateContent` returns:
> 
> ```
> const updated = await workflowService.update(user, data, id, {
>   source: 'n8n-mcp',
>   collectPostSaveError: true,  // only MCP passes this
> });
> // returns { workflow, postSaveError?: Error } shape for opted-in callers
> 
> // THEN ON THE TOOL SIDE: if (postSaveError) return successWithNote(...)
> ```
> 
> I know this will not solve the create tool but, if sound, it will establish a pattern that we can retrofit there. Let me know what you think.

Agreed! that's the right long-term direction. The service knows exactly where the commit boundary is, so moving post-save error capture there eliminates all the heuristic guessing in the tools.

Another approach can be a dedicated method rather than an optional flag, something like:

`updateWithPostSaveErrorCollection(user, workflowUpdateData, workflowId, options?)` returning `{ workflow, postSaveError?: Error }`.

The implementation would wrap the existing `update()` body: run pre-save validation and DB commit as usual, then catch only the post-commit steps (`workflow.afterUpdate`, `workflow-saved` event, and activation) separately. If any of those throw, return the already-persisted workflow plus the caught error; otherwise return just the workflow.

That keeps `update()` completely unchanged for all existing callers, avoids overloads, and gives the MCP tool a simple:

`if (postSaveError) return successWithNote(workflow, postSaveError)`.

Later we can add a similar `createWorkflowWithPostSaveErrorCollection()` for the create path.


Thanks!

**Review by MiloradFilipovic — DISMISSED** [review]
Looking good. I'll open up a follow-up ticket for suggested alternative approach that we'll handle internally.

**Comment by sinehypernova-0718** [comment]
Thanks for the approval, @MiloradFilipovic! Glad it’s in good shape now.

And really appreciate you considering the service‑layer approach, looking forward to the follow‑up ticket. Thanks again for the review! ☺️

EDIT: Noticed the merge conflicts, just some imports. I have resolved the conflicts, you'd need to reapprove.

**Review by MiloradFilipovic — DISMISSED** [review]

**Comment by MiloradFilipovic** [comment]
@sinehypernova-0718 Feel free to merge this when ready

**Comment by sinehypernova-0718** [comment]
Thanks @MiloradFilipovic! All good from my side too. We can merge it now, please trigger the CI checks and then it's GTG. ✅

**Comment by sinehypernova-0718** [comment]
@MiloradFilipovic the branch some merge conflicts again. Once again it was an import (I should probably stop fixing the import order in my PRs 😅). You'd need to reapprove again :(

**Comment by sinehypernova-0718** [comment]
@MiloradFilipovic Apologies for the delay. While resolving the merge conflicts, I by mistake removed an import causing the build failures. The fix is pushed now, Build and Typecheck are green locally. Can you please re-trigger the CI checks once you get a chance?

Thanks!

**Review by MiloradFilipovic — DISMISSED** [review]

**Comment by MiloradFilipovic** [comment]
> @MiloradFilipovic Apologies for the delay. While resolving the merge conflicts, I by mistake removed an import causing the build failures. The fix is pushed now, Build and Typecheck are green locally. Can you please re-trigger the CI checks once you get a chance?
> 
> Thanks!

Alright, just ran those, looks like there is [a format issue](https://github.com/n8n-io/n8n/actions/runs/34614192056/job/103932704157?pr=34975). Can you check please?

**Comment by sinehypernova-0718** [comment]
> > @MiloradFilipovic Apologies for the delay. While resolving the merge conflicts, I by mistake removed an import causing the build failures. The fix is pushed now, Build and Typecheck are green locally. Can you please re-trigger the CI checks once you get a chance?
> > Thanks!
> 
> Alright, just ran those, looks like there is [a format issue](https://github.com/n8n-io/n8n/actions/runs/34614192056/job/103932704157?pr=34975). Can you check please?

The fix is pushed now.

**Comment by sinehypernova-0718** [comment]
Hey @MiloradFilipovic! sorry this took longer than it should have.

Everything is pushed now. Lint, build, and tests are all green on the latest commit. The only failing check is [shard 2](https://github.com/n8n-io/n8n/actions/runs/34847113930/job/103990532186?pr=34975) of the e2e suite. It looks unrelated to my changes, but I'm not fully sure, so if you can confirm, that'd help.

One note on the lint fix: I had to revert `deepCopy` back to a JSON round-trip for the `normalize` helper. The DB drops `undefined` properties on round-trip, and `deepCopy` preserves them, so using it broke recovery for any update that set a parameter to `undefined`. Since we want the shape the DB would have stored, the JSON semantics are actually the correct behaviour here. There's an inline `eslint-disable` for `no-json-parse-json-stringify` with a comment explaining why.

Also, since the previous approval was dismissed by the new pushes, you'd need to reapprove for the required reviews workflow to pass, then we should be good to go. Thanks!

**Comment by scdekov** [comment]
Hi @sinehypernova-0718 can you merge latest master to this branch once again as I see there are new merge conflicts. I'll kick off the CI after that

**Comment by sinehypernova-0718** [comment]
@scdekov The conflicts have been resolved. Please trigger the CI once again. 🙏

**Comment by sinehypernova-0718** [comment]
Hey! All the checks have passed. It's ready for merge now, go ahead when you're ready. 

Thanks for the review!

**Review by scdekov — APPROVED** [review]
